### PR TITLE
Add Huggingface Token classification processor

### PR DESCRIPTION
**Pull request by jzpang**
This PR fixes https://github.com/asyml/forte-wrappers/issues/58. 

### Description of changes
Add Huggingface Token classification processor and test cases.

### Possible influences of this PR.
Describe what are the possible side-effects of the code change.

### Test Conducted
Describe what test cases are included for the PR.

### CODE REVIEW

File: src/huggingface/forte/huggingface/token_classification.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+
+class TokenClassification(PackProcessor):
+    r"""Wrapper of the models on HuggingFace platform with pipeline tag of
+    `ner`.
+    https://huggingface.co/models?pipeline_tag=token-classification
+    This wrapper could take any model name on HuggingFace platform with pipeline
+    tag of `ner` in configs to make prediction on the user
+    specified entry type in the input pack and the prediction result goes to the
+    user specified attribute name of that entry type in the output pack. User
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:42]: This docstring explanation is hard to follow. Can you explain what it does in simpler language? You can postpone the explanation of how to use attribute names and entry types, and focus on a high-level explanation first.

You can provide examples instead of using plain text to explain the configs (use a code block here). And then you can refer readers to the `default_config` function for more information. 

Put yourself into the user's hat, see if you can understand the usage of this by reading the docstring?

File: src/huggingface/forte/huggingface/token_classification.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    could provide task and models in the config to perform NER, POS,
+    word segmentation task, etc.
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:44]: Please try to grammar-check the sentences as well. For example, "Users could provide tasks and models" or "A user could provide tasks and models"

File: src/huggingface/forte/huggingface/token_classification.py [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.classifier = None
+
+    def set_up(self):
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:52]: Do we need a method here? This could simply be in the `initialize` method. At least this shouldn't be a public method.

File: src/huggingface/forte/huggingface/token_classification.py [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        mod = importlib.import_module(input_path)
+        input_entry = getattr(mod, input_module)
+
+        output_path, output_module = self.configs.output_entry_type.rsplit(
+            ".", 1
+        )
+        mod = importlib.import_module(output_path)
+        output_entry = getattr(mod, output_module)
+
+        for entry_specified in input_pack.get(entry_type=input_entry):
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:82]: The `get` function supports to get entry from strings directly. You don't have to deal with the modules and classes. 
```suggestion
        for entry_specified in input_pack.get(entry_type= self.configs.entry_type):
```

File: src/huggingface/forte/huggingface/token_classification.py [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            tokenizer=self.configs.tokenizer,
+            framework=self.configs.framework,
+            device=self.configs.cuda_device,
+        )
+
+    def initialize(self, resources: Resources, configs: Config):
+        super().initialize(resources, configs)
+        self.set_up()
+
+    def _process(self, input_pack: DataPack):
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:65]: Missing the return type in the function signature.

File: src/huggingface/forte/huggingface/token_classification.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        output_path, output_module = self.configs.output_entry_type.rsplit(
+            ".", 1
+        )
+        mod = importlib.import_module(output_path)
+        output_entry = getattr(mod, output_module)
+
+        for entry_specified in input_pack.get(entry_type=input_entry):
+            result = self.classifier(entry_specified.text)
+
+            if self.configs.strategy == "bio-merge":  # Merge BIO tagging
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:85]: `strategy` does not explain what this does. Maybe call it `tagging_scheme`?

File: src/huggingface/forte/huggingface/token_classification.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+                result_types, result_indices = self._merge_bio_tokens(result)
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:86]: Let's label some variable types to make mypy useful.

File: src/huggingface/forte/huggingface/token_classification.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+            else:
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:88]: 1. We don't have the test case for this branch. 
1. Try not to use `else` directly, use `elif` and consider the `else` branch as InvalidConfig

File: src/huggingface/forte/huggingface/token_classification.py [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+        """
+        input_path, input_module = self.configs.entry_type.rsplit(".", 1)
+        mod = importlib.import_module(input_path)
+        input_entry = getattr(mod, input_module)
+
+        output_path, output_module = self.configs.output_entry_type.rsplit(
+            ".", 1
+        )
+        mod = importlib.import_module(output_path)
+        output_entry = getattr(mod, output_module)
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:80]: Please reuse the library for getting class ("https://github.com/asyml/forte/blob/master/forte/utils/utils.py#L77")

Here is an example: https://github.com/asyml/forte/blob/master/forte/processors/misc/delete_overlap_entries.py#L46

File: src/huggingface/forte/huggingface/token_classification.py [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+                    begin=entry_specified.span.begin + int(start),
+                    end=entry_specified.span.begin + int(end),
+                )
+
+                if self.configs.task == "ner":
+                    if output_module == "EntityMention":
+                        entity.ner_type = type
+                    else:
+                        try:
+                            entity.ner = type
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:108]: Why `entity.ner_type` at line 105 but `entity.ner` here? I suppose you should assign the attribute based on the attribute name, like this:

https://github.com/asyml/forte-wrappers/blob/main/src/huggingface/forte/huggingface/zero_shot_classifier.py#L76

Talk to Jenny if you are not sure about this.

File: src/huggingface/forte/huggingface/token_classification.py [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+
+            for type, (start, end) in zip(result_types, result_indices):
+                entity = output_entry(
+                    pack=input_pack,
+                    begin=entry_specified.span.begin + int(start),
+                    end=entry_specified.span.begin + int(end),
+                )
+
+                if self.configs.task == "ner":
+                    if output_module == "EntityMention":
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:104]: We shouldn't hardcode the output module such as `EntityMention`, use the setattr idea explained below

File: src/huggingface/forte/huggingface/token_classification.py [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+                    else:
+                        try:
+                            entity.ner = type
+                        except KeyError as exc:
+                            raise ProcessorConfigError(
+                                f"NER type was not stored in the given "
+                                f"'output_entry_type': "
+                                f"{self.configs.output_entry_type}."
+                                f"EntityMention or Token was recommended."
+                            ) from exc
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:115]: We can support any entry type, so please don't hard code `EntityMention` and `Token` here.

File: src/huggingface/forte/huggingface/token_classification.py [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+                if self.configs.task == "pos":
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:117]: Same as above, don't hard code attribute name.

File: src/huggingface/forte/huggingface/token_classification.py [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+                        entity.pos = type
+                    except KeyError as exc:
+                        raise ProcessorConfigError(
+                            f"POS type was not stored in the given "
+                            f"'output_entry_type': "
+                            f"{self.configs.output_entry_type}."
+                            f"Token was recommended."
+                        ) from exc
+
+    def _merge_bio_tokens(self, tokens):
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:128]: Add parameter type and return type.

File: src/huggingface/forte/huggingface/token_classification.py [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+            - `task`: defines the nlp task, for example, "ner", "pos", "ws".
+            - `strategy`: defines the result merged strategy, for example,
+              "no_merge" to output the original result, "bio-merge" to merge
+              the BIO tagging format result. The default is "no_merge".
+            - `model_name`: language model, default is `"dslim/bert-base-NER"`.
+              The wrapper supports HuggingFace models with pipeline tag of
+              `ner`.
+            - `tokenizer`: language model for tokenization, default is
+              `"dslim/bert-base-NER"`.
+            - `framework`: The framework of the model, should be `"pt"` or
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:176]: Please label if this information should be obtained from Huggingface model hub.

File: src/huggingface/forte/huggingface/token_classification.py [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+              `"tf"`.
+            - `cuda_device`: Device ordinal for CPU/GPU supports. Setting
+              this to -1 will leverage CPU, a positive will run the model
+              on the associated CUDA device id.
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:180]: default is -1

File: src/huggingface/forte/huggingface/token_classification.py [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+        would be checked before running the processor if
+        the pipeline is initialized with
+        `enforce_consistency=True` or
+        :meth:`~forte.pipeline.Pipeline.enforce_consistency` was enabled for
+        the pipeline.
+        """
+        return {cls().default_configs()["entry_type"]: set()}
+
+    def record(self, record_meta: Dict[str, Set[str]]):
+        r"""Method to add output type record of `ner` which is
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:211]: Output record may not be `ner` right?

**Review by hunterhector — COMMENTED** [review]
See comments

File: src/huggingface/forte/huggingface/token_classification.py [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+__all__ = [
+    "TokenClassification",
+]
+
+
+class TokenClassification(PackProcessor):
+    r"""Wrapper of the models on HuggingFace platform with pipeline tag of
+    `ner`.
+    https://huggingface.co/models?pipeline_tag=token-classification
+    This wrapper could take any model name on HuggingFace platform with pipeline
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:39]: It might be a bit exaggerated to say "any"

File: src/huggingface/forte/huggingface/token_classification.py (hunk 5 above)
Comment (jzpang) [reply src/huggingface/forte/huggingface/token_classification.py:65]: _process() has no return in this processor

File: src/huggingface/forte/huggingface/token_classification.py [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+    tag of `ner` in configs to make prediction on the user
```
[the remark this answers is not in the document]
Comment (jzpang) [reply src/huggingface/forte/huggingface/token_classification.py]: Right, this is the tag used in Huggingface model hub. It is fixed.

File: src/huggingface/forte/huggingface/token_classification.py [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+    `ner`.
+    https://huggingface.co/models?pipeline_tag=token-classification
+    This wrapper could take the models on HuggingFace platform with pipeline
+    tag of `ner` to make prediction on token classification tasks, for example,
+    NER, POS tagging, word segmentation, etc.
+    This processor takes user specified entry type in the input pack and the
+    prediction result goes to the user specified attribute name of that entry
+    type in the data pack. Please refer to the `default_config` function for
+    more information.
+
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:46]: Let's provide examples in code block to explain how to use NER, WS.

File: src/huggingface/forte/huggingface/token_classification.py [hunk 21]
```diff
@@ -0,0 +1,204 @@
...
+            input_pack: Input pack to fill
+        Returns:
+        """
+        if not self.configs.entry_type:
+            raise ProcessorConfigError("Please specify an input entry type!")
+
+        input_entry = get_class(self.configs.entry_type)
+        output_entry = get_class(self.configs.output_entry_type)
+
+        for entry_specified in input_pack.get(entry_type=input_entry):
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:76]: We don't need to get the class first, we can use the class name (str) to get entries.

```suggestion
        for entry_specified in input_pack.get(entry_type= self.configs.entry_type):
```

File: src/huggingface/forte/huggingface/token_classification.py [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+              on the associated CUDA device id. Default is -1.
+
+        Returns: A dictionary with the default config for this processor.
+        """
+        config = super().default_configs()
+        config.update(
+            {
+                "entry_type": "ft.onto.base_ontology.Sentence",
+                "output_entry_type": "ft.onto.base_ontology.Token",
+                "task": "ner",  # "pos", "ws"
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:172]: I think the `task` config is not useful anymore.

File: src/huggingface/forte/huggingface/token_classification.py [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+              this to -1 will leverage CPU, a positive will run the model
+              on the associated CUDA device id. Default is -1.
+
+        Returns: A dictionary with the default config for this processor.
+        """
+        config = super().default_configs()
+        config.update(
+            {
+                "entry_type": "ft.onto.base_ontology.Sentence",
+                "output_entry_type": "ft.onto.base_ontology.Token",
```
Comment (hunterhector) [inline src/huggingface/forte/huggingface/token_classification.py:171]: If we want to use NER as default, let's not use `Token.ner` to store them, let's make the default to be `EntityMention`, and `tagging-schem` to be `bio-merge`.

File: tests/wrappers/huggingface/token_classification_classifier_test.py [hunk 24]
```diff
@@ -0,0 +1,250 @@
...
+from forte.data.readers import StringReader
+from forte.nltk import NLTKSentenceSegmenter
+from forte.huggingface import TokenClassification
+
+
+class TestTokenClassification(unittest.TestCase):
+    def test_huggingface_ner_bio_classification(self):
+        nlp = Pipeline[DataPack](enforce_consistency=True)
+        nlp.set_reader(StringReader())
+        nlp.add(NLTKSentenceSegmenter())
```
Comment (hunterhector) [inline tests/wrappers/huggingface/token_classification_classifier_test.py:30]: We can reduce the dependency to NLTK by using this sentence splitter based on periods (https://github.com/asyml/forte/blob/master/forte/processors/misc/simple_processors.py#L26)

**Review by hunterhector — COMMENTED** [review]
Test cases look good, only some minor comments.

File: tests/wrappers/huggingface/token_classification_classifier_test.py (hunk 24 above)
Comment (jzpang) [reply tests/wrappers/huggingface/token_classification_classifier_test.py:30]: Using this processor will lead a error on pipeline:
forte.common.exception.ExpectedRecordNotFound: The record type ft.onto.base_ontology.Sentence is not found in meta of the prediction datapack

So I continued with NLTK.

Comment (hunterhector) [reply tests/wrappers/huggingface/token_classification_classifier_test.py:30]: Remove the `enforce_consistency` argument in the pipeline, which will result in the pipeline not checking the expected types.

btw, next time, please do not resolve the conversation so that I can see your replies (otherwise they will be collapsed).__

File: .github/workflows/main.yml [hunk 25]
```diff
@@ -87,7 +87,7 @@ jobs:
           - { dep: "nltk gpt2", extra: "termcolor>=1.1.0", testfile: examples/gpt2_test.py }
           - { dep: elastic, testfile: tests/wrappers/elastic_indexers_test.py }
           - { dep: faiss, testfile: tests/wrappers/faiss_indexers_test.py }
-          - { dep: "huggingface nltk", testfile: tests/wrappers/huggingface }
+          - { dep: "huggingface nltk", extra: "tensorflow>=2.5.0", testfile: tests/wrappers/huggingface }
```
Comment (hunterhector) [inline .github/workflows/main.yml:90]: Is `nltk` still needed here?

Comment (jzpang) [reply .github/workflows/main.yml:90]: Yes, I think one of Jenny's unittest need it

Comment (hunterhector) [reply .github/workflows/main.yml:90]: i see, missed it on that PR. Will approve this.

**Review by hunterhector — APPROVED** [review]
